feat: add authentication related feature

thdefn opened the pull request:
## 🔎 PR Description
- Close #7 
> add authentication related feature using spring security

## 🔑 Key Changes
- add login feature
- add register feature
- add check email feature
- set up token based authentication using spring security

## 📢 To Reviewers
- Detailed jwt exception could not handle yet. It seems like that should be done after communicating with the FE.

thdefn committed:
feat: set up spring security

thdefn committed:
feat: apply token based authorization

thdefn committed:
feat: add join related api endpoint
- add register feature
- add login feature
- add usable email check feature

thdefn committed:
refactor: move security-related class to a common module for reusability

thdefn committed:
test: test that authentication related features works well

thdefn committed:
feat: handle exception during the authentication process

thdefn committed:
chore: correct typo

Rekite commented on gildongmu-api/src/main/java/codeit/api/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,56 @@
+    @ApiResponse
+    @io.swagger.v3.oas.annotations.parameters.RequestBody(content = @Content(encoding = @Encoding(name = "request", contentType = "application/json")))
+    @PostMapping(value = "/signup", consumes = MediaType.MULTIPART_FORM_DATA_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    private ResponseEntity<Void> register(@Valid @RequestPart SignUpRequest request,
```
혹시 public이 아닌 private로 설정하신 이유가 있을까요!?

thdefn replied:
저는 컨트롤러 클래스 메서드는 외부에서 사용할 일이 없으니까 웬만하면 private 으로 해두는 편이에요 !

Rekite commented on gildongmu-api/src/main/java/codeit/api/auth/dto/request/LogInRequest.java:
```diff
@@ -0,0 +1,23 @@
+    private String email;
+
+    @NotBlank(message = "invalid password")
+    @Size(min = 8, message = "invalid password")
```
blank와 다르게 사이즈 제한이니까 조금 더 의미있는 메시지는 어떨까요?

thdefn replied:
완 전 좋 습 니 다

Rekite commented on gildongmu-api/src/main/java/codeit/api/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,56 @@
+    private ResponseEntity<Void> register(@Valid @RequestPart SignUpRequest request,
+        @Valid @RequestPart(required = false) MultipartFile profile) {
+        authService.register(request, profile);
+        return ResponseEntity.ok().build();
```
이미 고려해보셨겠지만 http 응답코드 201도 한번 고려해보는 것은 어떨까요..?!

thdefn replied:
쪼아요 201이 확실히 더 의미가 명확할 것 같습니다 !

Rekite reviewed (commented):
LGTM

thdefn committed:
chore: modify validation error message more clearly
- modify validation error message
- modify http status code when register is successful
